src/pow.cpp

```diff
@@ -61,7 +61,19 @@ unsigned int CalculateNextWorkRequired(const CBlockIndex* pindexLast, int64_t nF
     // Retarget
     const arith_uint256 bnPowLimit = UintToArith256(params.powLimit);
     arith_uint256 bnNew;
-    bnNew.SetCompact(pindexLast->nBits);
+
+    // Special difficulty rule for Testnet4
+    if (params.fPowAllowMinDifficultyBlocks && params.hashGenesisBlock == uint256S("0x00000000da84f2bafbbc53dee25a72ae507ff4914b867c565be350b0da8bf043")) {
+        // Use the last non-special-min-difficulty-rules-block
+        const CBlockIndex* pindex = pindexLast;
+        const unsigned int pow_min{bnPowLimit.GetCompact()};
+        while (pindex->pprev && pindex->nHeight % params.DifficultyAdjustmentInterval() != 0 && pindex->nBits == pow_min)
```

Comment:
e172a96c0781de2bbd69312905d2c16cc1745c2f: this had me confused again, so maybe a comment is helpful. 

If miners go away in the middle of difficulty adjustment period A, this rule will find the last "real" `nBits` value and apply it to the first block of period B. If nothing changes, then for period C this seeks back to the first block of period B, finds the real value and applies it.

But then how do the first blocks of B and C ever mined?

Well, remember that `GetNextWorkRequired` ignores the `nBit` value if enough time went by.

A test would be nice, but unfortunately it seems the only way to do that is to introduce a new regtest with the same rule, but at much lower difficulty.

Reply:
> Well, remember that GetNextWorkRequired ignores the nBit value if enough time went by.

That last part isn't true. The first block of the difficulty period does not allow the usage of the 20-min rule. That code is wrapped in the following if statemnt so `GetNextWorkRequired` always passed off to `CalculateNextWorkRequired` for the first block in a new difficulty period.

```
// Only change once per difficulty adjustment interval
if ((pindexLast->nHeight+1) % params.DifficultyAdjustmentInterval() != 0) {
    ...
}
```

I managed to confuse myself with this again today as well, so I added an adapted version of your comment here. The confusing part for me is that this is not a change of behavior which is why I also failed to emphasize this in the BIP so far. The first block of the difficulty period also can not use the 20-min rule in Testnet 3. But there it can get difficulty 1 when the last block of the previous period was mined with the 20-min rule, which then leads to a block storm.

For this to work the way you describe we would need to add the 20-min also in `CalculateNextWorkRequired` but that just gives us back the good ol' block storms. Maybe we could add something crazy there like 24 hours, that would mean we reset to 1 to get unstuck if a block isn't found for a day without anyone pointing their asic to it. But I am not sure we really need this.

Reply:
As mentioned on the BIP PR, it would be easier to simply read the difficulty from the first block in each difficulty period since it must have the actual difficulty.

Reply:
Done, I am now using the first block